### ISSUE TITLE
Panic due to probable race condition acquiring pool conn when database restarts

### ISSUE THREAD

**Issue by niallnsec** [jackc/pgx#1419]
**Describe the bug**
I am seeing a panic occur in certain conditions when the database restarts. I have a piece of code which listens on a connection acquired from a pgxpool.Pool for notifications. In the event of a server restart, I get the following panic in pgx:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x7f8538]

goroutine 44 [running]:
github.com/jackc/puddle/v2.(*Resource[...]).Value(0x4000579e48?)
    /go/pkg/mod/github.com/jackc/puddle/v2@v2.1.2/pool.go:47 +0x18
github.com/jackc/pgx/v5/pgxpool.(*Conn).connResource(...)
    /go/pkg/mod/github.com/jackc/pgx/v5@v5.2.0/pgxpool/conn.go:121
github.com/jackc/pgx/v5/pgxpool.(*Conn).Conn(...)
    /go/pkg/mod/github.com/jackc/pgx/v5@v5.2.0/pgxpool/conn.go:117
github.com/jackc/pgx/v5/pgxpool.(*Conn).Exec(0x4000714b50?, {0x1511dc0, 0x40001184e0}, {0x4000be1bf0, 0xe}, {0x0, 0x0, 0x0})
    /go/pkg/mod/github.com/jackc/pgx/v5@v5.2.0/pgxpool/conn.go:83 +0x48
private.repo/api.(*dbPool).listen(0x1511d50?, {0x1511dc0, 0x40001184e0}, {0x40005102d0, 0x5, 0x5})
    /src/db.go:218 +0xb8
private.repo/api.(*notificationManager).mustListen(0x40006ac000, {0x1511d50, 0x40005582c0})
    /src/notify.go:36 +0xd0
...
```

The code doing the listening looks like this:
```go
func (dc *dbPool) mustAcquire(ctx context.Context) (conn *pgxpool.Conn, err error) {
	var tries int
	for {
		conn, err = dc.Pool.Acquire(ctx)
		if err == nil {
			return
		}

		observeDBAcquireError()

		if ctx.Err() != nil {
			err = ctx.Err()
			return
		}
		time.Sleep(defBackoff.Backoff(tries))
		tries++
	}
}

func (dc *dbPool) listen(ctx context.Context, channel ...string) (*dbListener, error) {
	for {
		conn, err := dc.mustAcquire(ctx)
		if err != nil {
			return nil, err
		}

		for i := 0; i < len(channel); i++ {
			_, err = conn.Exec(ctx, "LISTEN "+channel[i]) <--- crash occurs on this line
			if err != nil {
				if conn.Conn().IsClosed() {
					conn.Release()
					continue
				}

				_ = conn.Conn().Close(context.Background())

				conn.Release()

				return nil, err
			}
		}

		return &dbListener{
			conn:     conn,
			channels: channel,
		}, nil
	}
}
```

**Expected behavior**
I would expect this code to loop until it acquires a new conn and successfully starts listening again.

**Actual behavior**
A panic occurs. After a lot of trial and error I managed to observe that `conn, err = dc.Pool.Acquire(ctx)` (dc.Pool is a *pgxpool.Pool) was returning no error and a conn where the unexported `res` field is nil. As a result, when I try to use the conn on the line `_, err = conn.Exec(ctx, "LISTEN "+channel[i])` there is a panic.

Oddly, when debugging I was inserting fmt.Println statements and after a number were added the panic would stop triggering, which is what makes me think this is likely a race condition somewhere.

**Version**
 - Go: `go version go1.19.3 darwin/arm64`
 - PostgreSQL: `PostgreSQL 14.4 on aarch64-unknown-linux-musl, compiled by gcc (Alpine 11.2.1_git20220219) 11.2.1 20220219, 64-bit`
 - pgx: `v5.2.0` I have just upgraded to v5 in hopes of fixing this bug, the same thing was happening in v4.

**Comment by jackc** [comment]
I presume this should be `!=`?

```go
		conn, err = dc.Pool.Acquire(ctx)
		if err == nil {
			return
		}
```

**Comment by niallnsec** [comment]
No, conn and err are both named return params, so if err == nil then I want to return the acquired connection, otherwise I continue looping to try and get a valid connection.

**Comment by jackc** [comment]
> No, conn and err are both named return params, so if err == nil then I want to return the acquired connection, otherwise I continue looping to try and get a valid connection.

Ah, you're right, I didn't catch the bare return.

---

There's been a lot of changes between v4 and v5.2 with regard to the connection pool. It's very surprising that the same problem would surface the same way in both.

I've looked at your code and at the underlying pgx code and nothing jumps out at me.

Would you be able to provide a standalone program that demonstrates this?

**Comment by niallnsec** [comment]
I'm really sorry to have wasted your time, this is totally my bug! I can't believe I never spotted it before, but in the `listen` function above I am doing this:

```go
if conn.Conn().IsClosed() {
	conn.Release()
	continue
}
```

which is logically wrong. I intended to continue the *outer* loop, but instead I continue the inner loop. And since I am checking to see if conn is closed this error path happens very rarely, with a DB restart being the most common cause. So if `len(channel) > 1` then I end up reusing a conn which I just released, hence the subsequent crash.

Thanks for your help and sorry again!
